sql/database.go

```diff
@@ -0,0 +1,239 @@
+package sql
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"crawshaw.io/sqlite"
+	"crawshaw.io/sqlite/sqlitex"
+
+	"github.com/spacemeshos/go-spacemesh/database"
+)
+
+var (
+	// ErrNoConnection is returned if pooled connection is not available.
+	ErrNoConnection = errors.New("database: no free connection")
+	// ErrNotFound is returned if requested record is not found.
+	// TODO(dshulyak) is an alias to datatabase.ErrNotFound until full transition is implemented.
+	ErrNotFound = database.ErrNotFound
+	// ErrObjectExists is returned if database contraints didn't allow to insert an object.
+	ErrObjectExists = errors.New("database: object exists")
+)
+
+// Executor is an interface for executing raw statement.
+type Executor interface {
+	Exec(string, Encoder, Decoder) (int, error)
+}
+
+// Statement is an sqlite statement.
+type Statement = sqlite.Stmt
+
+// Encoder for parameters.
+// Both positional parameters:
+// select block from blocks where id = ?1;
+//
+// and named parameters are supported:
+// select blocks from blocks where id = @id;
+//
+// For complete information see https://www.sqlite.org/c3ref/bind_blob.html.
+type Encoder func(*Statement)
+
+// Decoder for sqlite rows.
+type Decoder func(*Statement) bool
+
+func defaultConf() *conf {
+	return &conf{
+		connections: 16,
+		migrations:  embeddedMigrations,
+	}
+}
+
+type conf struct {
+	flags       sqlite.OpenFlags
+	connections int
+	migrations  Migrations
+}
+
+// WithConnections overwrites number of pooled connections.
+func WithConnections(n int) Opt {
+	return func(c *conf) {
+		c.connections = n
+	}
+}
+
+// WithMigrations overwrites embedded migrations.
+func WithMigrations(migrations Migrations) Opt {
+	return func(c *conf) {
+		c.migrations = migrations
+	}
+}
+
+// Opt for configuring database.
+type Opt func(c *conf)
+
+// InMemory database for testing.
+func InMemory(opts ...Opt) *Database {
+	opts = append(opts, WithConnections(1))
+	db, err := Open("file::memory:?mode=memory", opts...)
+	if err != nil {
+		panic(err)
+	}
+	return db
+}
+
+// Open database with options.
+//
+// Database is opened in WAL mode and pragma synchronous=normal.
+// https://sqlite.org/wal.html
+// https://www.sqlite.org/pragma.html#pragma_synchronous
+func Open(uri string, opts ...Opt) (*Database, error) {
+	config := defaultConf()
+	for _, opt := range opts {
+		opt(config)
+	}
+	pool, err := sqlitex.Open(uri, config.flags, config.connections)
+	if err != nil {
+		return nil, fmt.Errorf("open db %s: %w", uri, err)
+	}
+	db := &Database{pool: pool}
+	if config.migrations != nil {
+		tx, err := db.Tx(context.Background())
+		if err != nil {
+			return nil, err
+		}
+		defer tx.Release()
+		if err := config.migrations(tx); err != nil {
+			return nil, err
+		}
+		tx.Commit()
+	}
+	return db, nil
+}
+
+// Database is an instance of sqlite database.
+type Database struct {
+	pool *sqlitex.Pool
+}
+
+// Tx creates deferred sqlite transaction.
```

Comment:
i am surprised that you need to write this wrappers in this file. what's the reason that the library doesn't provide them?

Reply:
i want db and tx to share common interface (Exec method), that's why it is implemented this way. otherwise, i don't find it surprising, as i mentioned this is a thin wrapper, it has some opinionated utilities for executing multiple statements atomically but they don't match what i need.